checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java

```diff
@@ -1,176 +1,850 @@
+    OWNING = AnnotationBuilder.fromClass(this.resourceLeakAtf.getElementUtils(), Owning.class);
+    methodTree = ((UnderlyingAST.CFGMethod) cfg.getUnderlyingAST()).getMethod();
+    methodElt = TreeUtils.elementFromDeclaration(methodTree);
+    classTree = TreePathUtil.enclosingClass(resourceLeakAtf.getPath(methodTree));
+    // elementFromDeclaration() returns null when no element exists for the class tree, which can
+    // happen for certain kinds of anonymous classes, such as PolyCollectorTypeVar.java in the
+    // all-systems test suite.
+    classElt = TreeUtils.elementFromDeclaration(classTree);
+
+    if (classElt != null) {
+      for (Element memberElt : classElt.getEnclosedElements()) {
+        if (memberElt.getKind().isField() && resourceLeakAtf.hasOwning(memberElt)) {
+          owningFields.add((VariableElement) memberElt);
+        }
+      }
+    }
   }
 
   /**
-   * Runs the inference algorithm on the contents of the {@link #cfg} field.
+   * Creates a MustCallInference instance and runs the inference algorithm. A type factory's
+   * postAnalyze method calls this, if Whole Program Inference is enabled.
    *
-   * <p>Operationally, it checks method invocations for fields with non-empty @MustCall obligations
-   * along all paths to the regular exit point in the method body of the method represented by
-   * {@link #cfg}, and updates the {@link #owningFields} set if it discovers an owning field whose
+   * @param resourceLeakAtf the type factory
+   * @param cfg the control flow graph of the method to check
+   * @param mcca the MustCallConsistencyAnalyzer
+   */
+  /*package-private*/ static void runMustCallInference(
+      ResourceLeakAnnotatedTypeFactory resourceLeakAtf,
+      ControlFlowGraph cfg,
+      MustCallConsistencyAnalyzer mcca) {
+    MustCallInference mustCallInferenceLogic = new MustCallInference(resourceLeakAtf, cfg, mcca);
+    mustCallInferenceLogic.runInference();
+  }
+
+  /**
+   * Runs the inference algorithm on the current method (the {@link #cfg} field). It updates the
+   * {@link #disposedFields} set or adds @Owning to the formal parameter if it discovers their
    * must-call obligations were satisfied along one of the checked paths.
+   *
+   * <p>Operationally, it checks method invocations for fields and parameters with
+   * non-empty @MustCall obligations along all paths to the regular exit point.
    */
-  /*package-private*/ void runInference() {
-    Set<Block> visited = new HashSet<>();
-    Deque<Block> worklist = new ArrayDeque<>();
-    Block entry = this.cfg.getEntryBlock();
+  private void runInference() {
 
-    worklist.add(entry);
-    visited.add(entry);
+    Set<BlockWithObligations> visited = new HashSet<>();
+    Deque<BlockWithObligations> worklist = new ArrayDeque<>();
+
+    {
+      BlockWithObligations entry =
+          new BlockWithObligations(cfg.getEntryBlock(), getNonEmptyMCParams(cfg));
+      worklist.add(entry);
+      visited.add(entry);
+    }
```

Comment:
Why is this code in a nested block?  Is it to scope the `entry` variable?  If so that seems like overkill and I would remove it.